### ISSUE TITLE
Passenger build fails against apache (-fPIC -pie)

### ISSUE THREAD

**Issue by sagarun** [phusion/passenger#1756]
Here is how CFLAGS returned by apxs2 looks like

```
$  apxs2 -q CFLAGS
-fPIC -pie -DDEFAULT_PIDLOG="logs/apache/apache.pid" -DDEFAULT_SCOREBOARD="logs/apache/apache.scoreboard" -DDEFAULT_LOCKFILE="logs/apache/apache.lock" -DDEFAULT_ERRORLOG="logs/apache/error" -DSERVER_CONFIG_FILE="conf/apache22/apache.conf" -DAP_UNSAFE_ERROR_LOG_UNESCAPED -DDEFAULT_SERVER_LIMIT=2048 -DNO_LINGCLOSE -g -O2 -Wall -pipe -MD -MP -DXYZ
```
So apache is built with -fPIC -pie.

Here is how passenger module build fails

```
c++ -shared buildout/apache2/mod_passenger.o buildout/apache2/Configuration.o buildout/apache2/Bucket.o buildout/apache2/Hooks.o -fPIC -o buildout/apache2/mod_passenger.so buildout/apache2/module_libpassenger_common/Logging.o buildout/apache2/module_libpassenger_common/Exceptions.o buildout/apache2/module_libpassenger_common/Utils/SystemTime.o buildout/apache2/module_libpassenger_common/Utils/StrIntUtils.o buildout/apache2/module_libpassenger_common/Utils/StrIntUtilsNoStrictAliasing.o buildout/apache2/module_libpassenger_common/Utils/IOUtils.o buildout/apache2/module_libpassenger_common/Utils.o buildout/apache2/module_libpassenger_common/vendor-modified/modp_b64.o buildout/apache2/module_libpassenger_common/AppTypes.o buildout/apache2/module_libboost_oxt.a -L/home/y/lib64 -Wl,-rpath,/home/y/lib64 -Wl,-rpath,/home/y/libexec64/apache22   -L/lib64 -fPIC -DLINUX -D_REENTRANT -I/home/y/include64/apr -I/home/y/include64/apr -fPIC -pie -DDEFAULT_PIDLOG="logs/apache/apache.pid" -DDEFAULT_SCOREBOARD="logs/apache/apache.scoreboard" -DDEFAULT_LOCKFILE="logs/apache/apache.lock" -DDEFAULT_ERRORLOG="logs/apache/error" -DSERVER_CONFIG_FILE="conf/apache22/apache.conf" -DAP_UNSAFE_ERROR_LOG_UNESCAPED -DDEFAULT_SERVER_LIMIT=2048 -DNO_LINGCLOSE -g -Wall -pipe -MD -MP -DXYZ -I/home/y/include64/apache22 -L/home/y/lib64 -lapr -L/home/y/lib64 -laprutil -lpthread -lrt
/usr/lib/gcc/x86_64-redhat-linux/4.4.7/../../../../lib64/Scrt1.o: In function `_start':
(.text+0x20): undefined reference to `main'
collect2: ld returned 1 exit status
```

**Comment by FooBarWidget** [comment]
Which OS and version is this? Are you compiling against a self-compiled Apache?

What happens if you remove the '-pie' option?

**Comment by sagarun** [comment]
OS is RHEL 6. Yes, i am compiling against a self compiled apache. If the -pie option isn't there everything is fine. Unfortunately i need to -pie for ASLR

**Comment by FooBarWidget** [comment]
-pie is only for executables. For libraries, the correct flag is -fPIC. -fPIC allows ASLR. So you need to ensure that your apxs doesn't output -pie.

**Comment by sagarun** [comment]
Yes you are right. -pie is for executable.  apache is an executable so -pie was added there, because of it apxs emits the same flag too. I was able to build other apache modules with -fPIC -pie. Only passenger module is failing. If passenger doesn't like -pie flag shouldn't you guys remove it in https://github.com/phusion/passenger/blob/stable-5.0/src/ruby_supportlib/phusion_passenger/platform_info/apache.rb#L582 ?
